File: src/problem1663/Solution1663.java
/*
 * 1663. Smallest String With A Given Numric Value
 *
 * - Difficulty: Medium
 * - Topics: String, Greedy
 * - Hint:
 *      1. Think greedily.
 *      2. If you build the string from the end to the beginning, it will
 *         always be optimal to put the highest possible character at the
 *         current index.
 */

public class Solution1663 {

    // private Map<Integer, String> alphabets;

    // public Q1663() {
    // this.generateDictAlphabets();
    // }

    // private void generateDictAlphabets() {
    // this.alphabets = new HashMap<Integer, String>();
    // alphabets.put(1, "a");
    // alphabets.put(2, "b");
    // alphabets.put(3, "c");
    // alphabets.put(4, "d");
    // alphabets.put(5, "e");
    // alphabets.put(6, "f");
    // alphabets.put(7, "g");
    // alphabets.put(8, "h");
    // alphabets.put(9, "i");
    // alphabets.put(10, "j");
    // alphabets.put(11, "k");
    // alphabets.put(12, "l");
    // alphabets.put(13, "m");
    // alphabets.put(14, "n");
    // alphabets.put(15, "o");
    // alphabets.put(16, "p");
    // alphabets.put(17, "q");
    // alphabets.put(18, "r");
    // alphabets.put(19, "s");
    // alphabets.put(20, "t");
    // alphabets.put(21, "u");
    // alphabets.put(22, "v");
    // alphabets.put(23, "w");
    // alphabets.put(24, "x");
    // alphabets.put(25, "y");
    // alphabets.put(26, "z");
    // }

    // public String toAlphabet(int x) {
    // return this.alphabets.get(x);
    // }

    // public String getSmallestString(int n, int k) {
    // String result = "";
    // int max = 26;

    // while (n > 0) {
    // int quotient = Integer.valueOf(k / max);
    // int remainder = k % max;

    // if (k > n) {
    // if (quotient > 1) {
    // if ((k - max) > (n - 1)) {
    // result = this.toAlphabet(max) + result;
    // k -= max;
    // } else {
    // k -= n - 1;
    // result = this.toAlphabet(k) + result;
    // }

    // } else if (quotient == 1) {
    // if (remainder >= (n - 1)) {
    // result = this.toAlphabet(max) + result;
    // k -= max;
    // max = remainder;
    // } else {
    // max -= (n - (1 + remainder));
    // result = this.toAlphabet(max) + result;
    // k -= max;
    // }

    // } else {
    // max = remainder - (n - 1);
    // result = this.toAlphabet(max) + result;
    // k -= max;
    // }

    // } else {
    // return "a".repeat(n) + result;
    // }

    // n--;

    // }

    // return result;
    // }

    public String toAlphabet(int x) {
        return String.valueOf((char) (96 + x));
    }

    /*
     * New approach
     * In case when n = 95474 and k = 2394483:
     * - decrease 91962 to 2 iterations; &
     * - decrease 2451 ms to to 671 ms (avg) runtimes
     */
    public String getSmallestString(int n, int k) {
        String result = "";
        int max = 26;

        while (n > 0) {
            if (k > n) {
                int quotient = Integer.valueOf(k / max);
                int remainder = k % max;

                if (quotient > 1) {
                    if ((k - max) > n) {
                        while ((k - max * quotient) < (n - quotient)) {
                            quotient--;
                        }

                        result = "z".repeat(quotient);
                        k -= max * quotient;
                        n -= quotient;
                        continue;
                    }

                    k -= (n - 1);
                    result = this.toAlphabet(k) + result;

                } else if (quotient == 1) {
                    if (remainder >= (n - 1)) {
                        result = this.toAlphabet(max) + result;
                        k -= max;
                        max = remainder;
                    } else {
                        max -= (n - (1 + remainder));
                        result = this.toAlphabet(max) + result;
                        k -= max;
                    }

                } else {
                    max = remainder - (n - 1);
                    result = this.toAlphabet(max) + result;
                    k -= max;
                }

            } else {
                return "a".repeat(n) + result;
            }

            n--;
        }

        return result;
    }

    public static void main(String[] args) {
        // int n = 3;
        // int k = 27;
        // expected: "aay"

        // int n = 5;
        // int k = 73;
        // expected: "aaszz"

        // int n = 95474;
        // int k = 2394483;
        // failed: time limit exceeded
        // [SOLVED]: new approach

        // int n = 504;
        // int k = 13000; (26*500 = 13000)
        // expected: 2 iteration

        // int n = 85;
        // int k = 2159;
        // expected:
        // "aayzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"

        // int n = 4;
        // int k = 100;
        // Runtime Error
        // java.lang.ArithmeticException: / by zero
        // [SOLVED]: replace for loops to while loops

        // int n = 80;
        // int k = 576;
        // step: 1.1.1 -> 1.1.2 -> 2
        // output:
        // "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa(zz)zzzzzzzzzzzzzzzzzzz"
        // expected:
        // "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa(av)zzzzzzzzzzzzzzzzzzz"

        // =====================================================

        // int n = 2;
        // int k = 2;
        // step: 2

        // int n = 2;
        // int k = 52;
        // step: 1.1.1

        // int n = 5;
        // int k = 54;
        // step: 1.1.1 -> 1.2.2 -> 2

        // int n = 5;
        // int k = 56;
        // step: 1.1.1 -> 1.3 -> 2

        // int n = 5;
        // int k = 55;
        // step: 1.1.1 -> 2

        // int n = 50;
        // int k = 55;
        // step: 1.1.2 -> 2

        // int n = 5;
        // int k = 50/51;
        // step: 1.2.1 -> 1.2.2 -> 2

        // int n = 3;
        // int k = 29/30;
        // step: 1.2.1 -> 1.2.2 -> 2

        // int n = 2;
        // int k = 30;
        // step: 1.2.1 -> 1.2.1

        // int n = 3;
        // int k = 27;
        // step: 1.2.2 -> 2

        // int n = 20;
        // int k = 27;
        // step: 1.2.2 -> 2

        int n = 95474;
        int k = 2394483;

        long start = System.currentTimeMillis();

        Solution1663 solution = new Solution1663();
        System.out.println(solution.getSmallestString(n, k));

        long end = System.currentTimeMillis();
        long execTime = end - start;
        System.out.println("\nExecution time: " + execTime + " ms.");
    }
}

File: src/problem1663/solution1663.py
"""
1663. Smallest String With A Given Numeric Value

- Difficulty: Medium
- Topics: String, Greedy
- Hint:
    1. Think greedily.
    2. If you build the string from the end to the beginning, it will
       always be optimal to put the highest possible character at the
       current index.
"""

import time


class Solution:
    # def to_alphabet(self, x: int) -> str:
    #     alphabet_values = {
    #         1: "a",
    #         2: "b",
    #         3: "c",
    #         4: "d",
    #         5: "e",
    #         6: "f",
    #         7: "g",
    #         8: "h",
    #         9: "i",
    #         10: "j",
    #         11: "k",
    #         12: "l",
    #         13: "m",
    #         14: "n",
    #         15: "o",
    #         16: "p",
    #         17: "q",
    #         18: "r",
    #         19: "s",
    #         20: "t",
    #         21: "u",
    #         22: "v",
    #         23: "w",
    #         24: "x",
    #         25: "y",
    #         26: "z",
    #     }
    #     return alphabet_values[x]

    # def getSmallestString(self, n: int, k: int) -> str:
    #     """
    #     :type n: int
    #     :type k: int
    #     :rtype: str
    #     """
    #     maximum = 26
    #     result = ""
    #     for _ in range(n):
    #         quotient = int(k / maximum)
    #         remainder = k % maximum

    #         if k > n:
    #             if quotient > 1:
    #                 if (k - maximum) > n - 1:
    #                     result = self.to_alphabet(maximum) + result
    #                     k -= maximum
    #                 else:
    #                     k -= n - 1
    #                     result = self.to_alphabet(k) + result
    #             elif quotient == 1:
    #                 if remainder >= n - 1:
    #                     result = self.to_alphabet(maximum) + result
    #                     k -= maximum
    #                     maximum = remainder
    #                 else:
    #                     temp = maximum - (n - (1 + remainder))
    #                     result = self.to_alphabet(temp) + result
    #                     k -= temp
    #                     maximum = maximum - temp
    #             else:
    #                 maximum = remainder - (n - 1)
    #                 result = self.to_alphabet(maximum) + result
    #                 k -= maximum
    #         else:
    #             result = n * "a" + result
    #             break

    #         n -= 1

    #     return result

    def getSmallestString(self, n: int, k: int) -> str:
        result = ""
        max_ = 26

        while n > 0:
            if k > n:
                quotient = int(k / max_)
                remainder = k % max_

                if quotient > 1:
                    if (k - max_) > n:
                        while (k - max_ * quotient) < (n - quotient):
                            quotient -= 1

                        result = "z" * quotient
                        k -= max_ * quotient
                        n -= quotient
                        continue

                    k -= n - 1
                    result = chr(96 + k) + result

                elif quotient == 1:
                    if remainder >= (n - 1):
                        result = chr(96 + max_) + result
                        k -= max_
                        max_ = remainder
                    else:
                        max_ -= n - (1 + remainder)
                        result = chr(96 + max_) + result
                        k -= max_

                else:
                    max_ = remainder - (n - 1)
                    result = chr(96 + max_) + result
                    k -= max_

            else:
                return "a" * n + result

            n -= 1

        return result


if __name__ == "__main__":
    # n_ = 3
    # k_ = 27
    # expected: "aay"

    # n_ = 5
    # k_ = 73
    # expected: "aaszz"

    # n = 85
    # k = 2159
    # expected:
    # 'aayzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz'

    n_ = 95474
    k_ = 2394483

    start = time.time()

    print(Solution().getSmallestString(n_, k_))

    end = time.time()
    exec_time = round((end - start) * 10**3)
    print(f"\nExecution time: {exec_time} ms.")
